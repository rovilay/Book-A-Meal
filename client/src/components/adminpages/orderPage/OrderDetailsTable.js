import React, { Component } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import classname from 'classnames';
import ReactPaginate from 'react-paginate';

import { orderDetailHead } from '../../../helpers/tableHeadData';
import OrderTableRow from './OrderTablerow';

class OrderDetailsTable extends Component {
  constructor(props) {
    super(props);

    this.handlePaginationClick = this.handlePaginationClick.bind(this);
  }

  /**
   * handles pagination changes
   *
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    const { getOrderMeals, orderedMealsPagination, content } = this.props;
    const { limit } = orderedMealsPagination;
    const { id: orderId } = content;
    const nextPage = data.selected;
    const newOffset = nextPage * limit;
    const mealsUrl = `/api/v1/orders/${orderId}/meals`;
    getOrderMeals(mealsUrl, { limit, offset: newOffset });
  }

  render() {
    const {
      title,
      content,
      orderedMealsPagination,
    } = this.props;

    const {
      User,
      Meals: orderMeals,
      deliveryAddress,
      createdAt: date
    } = content;

    const {
      offset,
      numOfPages,
      count
    } = orderedMealsPagination;

    let totalCost = 0;


    return (
      <div className="table-container">
        <div className="order-details">
          <h2 className="title">
            {title}
          </h2>
          <hr />
          <p>
            <span className="bold">Customer:</span> {
              `${User.firstName} ${User.lastName}`
            }
          </p>
          <p>
            <span className="bold">Date:</span> {moment(date).format('LL')}
          </p>
          <p>
            <span className="bold">Time:</span> {moment(date).format('LT')}
          </p>
          <p>
            <span className="bold">Address:</span> {deliveryAddress}
          </p>
        </div>

        <div className="container-test">
          <div className="row head">
            {
              orderDetailHead.map((head, i) => (
                <p
                  key={i}
                  className={classname('row-item', {
                    'meal-title': head === 'Meal' || head === 'orderId'
                  })}
                >
                  {head}
                </p>
              ))
            }
          </div>
          {
            orderMeals.map((meal, i) => {
              const {
                id,
                title: Meal,
                OrderMeal
              } = meal;
              const { portion, cost: UnitCost } = OrderMeal;
              const cost = UnitCost * portion;
              totalCost += cost;
              const item = {
                sn: ++i + offset,
                Meal,
                UnitCost,
                portion,
                cost
              };
              return (
                <OrderTableRow
                  key={i}
                  item={item}
                  id={id}
                  {...this.props}
                />
              );
            })
          }
        </div>
        <div className="total-cost">
          <span className="bold">Total Cost: </span> &#8358;{totalCost}
        </div>
        {
          // show only if meals or orderMeals present
          (count > 5 || orderMeals.length < 1)
          &&
          <div className="modal-pagination">
            <ReactPaginate
              previousLabel="<<"
              nextLabel=">>"
              breakLabel={<a href="">...</a>}
              breakClassName="break-me"
              pageCount={numOfPages}
              marginPagesDisplayed={2}
              pageRangeDisplayed={5}
              onPageChange={this.handlePaginationClick}
              containerClassName="pagination"
              subContainerClassName="pages pagination"
              activeClassName="active"
            />
          </div>
        }
      </div>
    );
  }
}

OrderDetailsTable.propTypes = {
  title: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired,
  orderedMealsPagination: PropTypes.object.isRequired,
  getOrderMeals: PropTypes.func.isRequired
};

export default OrderDetailsTable;
