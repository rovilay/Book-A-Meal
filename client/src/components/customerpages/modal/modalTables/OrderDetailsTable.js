import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import ReactPaginate from 'react-paginate';
import classname from 'classnames';

import { orderDetailHead } from '../../../../helpers/tableHeadData';
import CustomerOrderTableRow from '../../orderPage/CustomerOrderTablerow';

class OrderDetailsTable extends Component {
  constructor(props) {
    super(props);

    this.handlePaginationClick = this.handlePaginationClick.bind(this);
  }

  /**
   * handles pagination changes
   *
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    const { id: orderId } = this.props.content;
    const { limit } = this.props.orderedMealsPagination;
    const nextPage = data.selected;
    const newOffset = nextPage * limit;
    const mealsUrl = `/api/v1/orders/${orderId}/meals`;
    this.props.getOrderMeals(mealsUrl, { limit, offset: newOffset });
  }

  render() {
    const {
      title,
      content,
      orderedMealsPagination,
    } = this.props;

    const {
      Meals: orderMeals,
      deliveryAddress,
      createdAt,
    } = content;

    const {
      offset,
      numOfPages,
      count
    } = orderedMealsPagination;

    return (
      <div className="modal-container">
        <div className="order-details">
          <h2 className="title">
            {title}
          </h2>
          <hr />
          <p>
            <span className="bold">Date:</span> {moment(createdAt).format('LL')}
          </p>
          <p>
            <span className="bold">Time:</span> {moment(createdAt).format('HH:mm')}
          </p>
          <p>
            <span className="bold">Address:</span> {deliveryAddress}
          </p>
        </div>

        <div className="container-test">
          <div className="row head">
            {
                orderDetailHead.map((head, i) => (
                  <p
                    key={i}
                    className={classname('row-item', { 'meal-title': head === 'Meal' || head === 'orderId' })}
                  >
                    {head}
                  </p>
                ))
              }
          </div>
          {
            orderMeals.map((meal, i) => {
              const {
                id,
                title: Meal,
                OrderMeal
              } = meal;
              const { portion, cost: unitCost } = OrderMeal;
              const cost = unitCost * portion;
              const item = {
                sn: ++i + offset,
                Meal,
                unitCost,
                portion,
                cost
              };

              return (
                <CustomerOrderTableRow
                  key={i}
                  item={item}
                  id={id}
                  {...this.props}
                />
              );
            })
          }
        </div>
        {
          // show only if meals or orderMeals present
          (count > 5)
          &&
          <div className="modal-pagination">
            <ReactPaginate
              previousLabel="<<"
              nextLabel=">>"
              breakLabel={<a href="">...</a>}
              breakClassName="break-me"
              pageCount={numOfPages}
              marginPagesDisplayed={2}
              pageRangeDisplayed={5}
              onPageChange={this.handlePaginationClick}
              containerClassName="pagination"
              subContainerClassName="pages pagination"
              activeClassName="active"
            />
          </div>
        }
      </div>
    );
  }
}

OrderDetailsTable.propTypes = {
  title: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired,
  getOrderMeals: PropTypes.func.isRequired,
  orderedMealsPagination: PropTypes.object.isRequired
};

export default OrderDetailsTable;
