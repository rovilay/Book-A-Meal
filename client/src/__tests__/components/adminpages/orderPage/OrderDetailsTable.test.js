import React from 'react';
import { shallow } from 'enzyme';
import OrderDetailsTable from '../../../../components/adminpages/orderPage/OrderDetailsTable';
import { orderDetails } from '../../../__mockData__/orderMock';


const props = {
  getOrderMeals: jest.fn(),
  title: 'Order Details',
  content: {
    ...orderDetails,
    User: {
      firstName: 'john',
      lastName: 'doe'
    },
    createdAt: '2018-09-03'
  },
  orderedMealsPagination: {
    limit: 12,
    count: 13,
    numOfPages: 2
  },
};

describe('OrderDetailsTable component test', () => {
  const setup = () => shallow(<OrderDetailsTable {...props} />);


  it('should render `OrderDetailsTable` correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });


  it('should call `handlePaginationClick` if pagination button is clicked', (done) => {
    const wrapper = setup();

    const data = {
      selected: 1
    };

    const handlePaginationClickSpy= jest.spyOn(wrapper.instance(), 'handlePaginationClick');
    wrapper.instance().handlePaginationClick(data)
    expect(handlePaginationClickSpy).toHaveBeenCalled();

    done();
  });
});
