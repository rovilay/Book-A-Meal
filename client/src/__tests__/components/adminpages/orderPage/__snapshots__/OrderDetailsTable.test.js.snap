// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetailsTable component test should render \`OrderDetailsTable\` correctly 1`] = `
<div
  className="table-container"
>
  <div
    className="order-details"
  >
    <h2
      className="title"
    >
      Order Details
    </h2>
    <hr />
    <p>
      <span
        className="bold"
      >
        Customer:
      </span>
       
      john doe
    </p>
    <p>
      <span
        className="bold"
      >
        Date:
      </span>
       
      September 3, 2018
    </p>
    <p>
      <span
        className="bold"
      >
        Time:
      </span>
       
      12:00 AM
    </p>
    <p>
      <span
        className="bold"
      >
        Address:
      </span>
       
      isuti rd
    </p>
  </div>
  <div
    className="container-test"
  >
    <div
      className="row head"
    >
      <p
        className="row-item"
        key="0"
      >
        S/N
      </p>
      <p
        className="row-item meal-title"
        key="1"
      >
        Meal
      </p>
      <p
        className="row-item"
        key="2"
      >
        Unit Cost
      </p>
      <p
        className="row-item"
        key="3"
      >
        Portion
      </p>
      <p
        className="row-item"
        key="4"
      >
        Cost
      </p>
    </div>
    <OrderTableRow
      content={
        Object {
          "Meals": Array [
            Object {
              "OrderMeal": Object {
                "cost": 400,
                "portion": 1,
              },
              "id": 1,
              "title": "Eba and Egusi",
            },
            Object {
              "OrderMeal": Object {
                "cost": 800,
                "portion": 2,
              },
              "id": 2,
              "title": "Ewa",
            },
          ],
          "User": Object {
            "firstName": "john",
            "lastName": "doe",
          },
          "createdAt": "2018-09-03",
          "deliveryAddress": "isuti rd",
          "orderId": 6,
        }
      }
      getOrderMeals={[MockFunction]}
      id={1}
      item={
        Object {
          "Meal": "Eba and Egusi",
          "UnitCost": 400,
          "cost": 400,
          "portion": 1,
          "sn": NaN,
        }
      }
      key="1"
      orderedMealsPagination={
        Object {
          "count": 13,
          "limit": 12,
          "numOfPages": 2,
        }
      }
      title="Order Details"
    />
    <OrderTableRow
      content={
        Object {
          "Meals": Array [
            Object {
              "OrderMeal": Object {
                "cost": 400,
                "portion": 1,
              },
              "id": 1,
              "title": "Eba and Egusi",
            },
            Object {
              "OrderMeal": Object {
                "cost": 800,
                "portion": 2,
              },
              "id": 2,
              "title": "Ewa",
            },
          ],
          "User": Object {
            "firstName": "john",
            "lastName": "doe",
          },
          "createdAt": "2018-09-03",
          "deliveryAddress": "isuti rd",
          "orderId": 6,
        }
      }
      getOrderMeals={[MockFunction]}
      id={2}
      item={
        Object {
          "Meal": "Ewa",
          "UnitCost": 800,
          "cost": 1600,
          "portion": 2,
          "sn": NaN,
        }
      }
      key="2"
      orderedMealsPagination={
        Object {
          "count": 13,
          "limit": 12,
          "numOfPages": 2,
        }
      }
      title="Order Details"
    />
  </div>
  <div
    className="total-cost"
  >
    <span
      className="bold"
    >
      Total Cost: 
    </span>
     â‚¦
    2000
  </div>
  <div
    className="modal-pagination"
  >
    <PaginationBoxView
      activeClassName="active"
      breakClassName="break-me"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel=">>"
      onPageChange={[Function]}
      pageCount={2}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel="<<"
      subContainerClassName="pages pagination"
    />
  </div>
</div>
`;
